### DISCUSSION TITLE
Repository pattern with prisma

### DISCUSSION THREAD

**Discussion by igolka97** [prisma/prisma#10584] [Q&A] [answered]
Hello team!

I want to thank you for a great product that I happily use every day for all my projects! 

I study various architectural patterns and approaches to bring order to the levels of logic of my next app and simplify my life and future development.  I decided to put my business logic in the service layer and work with the database in the repository to encapsulate all database interaction behind the repository as a black box.

I found that the prisma is not a traditional orm so it doesn't fits well when designing Hexagonal-like architecture.

Firstly, why?
Secondly, what approaches and patterns do you recommend for encapsulating prisma behind a layer of abstraction?

Maybe I’m wrong, and I should think in terms that Prisma is repository itself?

Thanks in advance!

**Comment by austincrim** [comment, marked as answer]
Hey @igolka97 👋, great question.

While Prisma is unlike traditional ORMs in many ways, it is still possible to wrap it in a thin layer of abstraction if that is something you want for your architecture. Most of the time, wrapping Prisma looks like writing wrapper functions around Prisma queries instead of the usual class-based pattern.

This is a reasonably common question, so here are a few links to previous discussions and questions around this topic:

- https://github.com/prisma/prisma/discussions/3929
- [Slack conversation about repository pattern](https://prisma.slack.com/archives/CKQTGR6T0/p1602664466064800)
- https://github.com/prisma/prisma/issues/5273

I would be happy to address any more specific questions you have!

**Reply by showduhtung** [reply]
Hi! Is there any way to publish the slack conversation somewhere? I can't seem to join the slack channel, as it seems like the community has moved to discord?

**Reply by ludralph** [reply]
Hi @showduhtung 

This is the contant of the slack conversation below:

Hi Prisma !
We are currently using Objection / knex and considering switching to Prisma which looks great, since we love reliable typings (types in Objection are quite good, but far from flawless).
There are two (similar) patterns that we use widely (in every query actually), and I can't see how to implement similar patterns with Prisma.
What would be the Prisma way of doing something equivalent ?
```js
// For the context, a normal query in Objection
const user = await User.query()
  .where("company_id", 123456)
  .where("name", "Paul");

/*
 * Pattern 1 : reusing queries
 */

// Definition :

class User extends ObjectionModel {
  static fromCompany = (company_id: number) => {
    return User.query().where("company_id", company_id);
  };
}

// Usage :

const my_company_id = 123456;
const colleague = await User.fromCompany(my_company_id).where("name", "Paul");
// this is equivalent to the first query above
// this way I'm sure that I never access user from another company by mistake

/*
 * Pattern 2 : composing queries
 */

// Definition :

export const filterRecent = <T extends ObjectionModel>(
  q: QueryBuilder<T, T[]>
) => {
  const now = new Date();
  const aMonthAgo = subMonths(now, 1);
  return q.whereBetween("created_at", [aMonthAgo, now]);
};

// Usage :

const recentUsers = filterRecent(User.fromCompany(my_company_id));

// these two are equivalent
const recentPostsFromMe_1 = filterRecent(Post.query()).where("author_id", my_user_id);
const recentPostsFromMe_2 = filterRecent(Post.query().where("author_id", my_user_id));

// this way I can re-use common part of queries
```

Hey 
[@Domi](https://prisma.slack.com/team/U01BARWCFUP)
 :wave:
Prisma doesn’t have a class based model like Objection does, it wouldn’t be possible to use these patterns directly.
Assuming your schema.prisma is something like this:
```prisma
model User {
  id      Int     @id @default(autoincrement())
  name    String
  company Company
}

model Company {
  id     Int    @id @default(autoincrement())
  name   String
  user   User   @relation(fields: [userId], references: [id])
  userId Int
}
```
What you could do is create a function:
```js
function queryFromCompany(companyId: number, where: Prisma.UserWhereInput) {
  return prisma.user.findMany({
    where: { company: { id: companyId }, ...where },
  })
}
```
And then use it like this:
```js
const user = await queryFromCompany(1, { name: 'Paul' })
```
I would still need to check for the second use case.


Thanks for your quick answer 
I'm (more than) fine with not using classes, I prefer a more functional approach anyway.
I agree with you on the first use-case, I would have done something similar. This isn't the most readable piece of code compared to similar stuff on other ORMs, but I guess it's fine if it's encapsulated in a single place.
And indeed, how to implement the second pattern is less clear to me.
Thanks !

I can’t find any other way for the second pattern, apart from this. Any other method would also forego type-safety in the process so I’m not sure how to go about that.
```js
function fromCompany(companyId: number): Prisma.UserWhereInput {
  return { companyId }
}

function filterRecent(
  start: Prisma.DateTimeFilter['equals'],
  end: Prisma.DateTimeFilter['equals']
): Pick<Prisma.UserWhereInput, 'createdAt'> {
  return {
    createdAt: {
      gte: start,
      lte: end,
    },
  }
}

const users = await prisma.user.findMany({
    where: {
      ...fromCompany(companyId),
      ...filterRecent(startDate, endDate),
    },
  })
```
The same can be done for Posts as well and will retain the types as well in this case.


Thank you.
The usage looks fine and quite readable, maybe it's indeed the way to go !
